### PR TITLE
Move postings back to int[] to take advantage of having more lanes per vector.

### PR DESCRIPTION
**Pull request by jpountz**
In Lucene 8.4, we updated postings to work on long[] arrays internally. This allowed us to workaround the lack of explicit vectorization (auto-vectorization doesn't detect all the scenarios that we would like to handle) support in the JVM by summing up two integers in one operation for instance.

With explicit vectorization now available, it looks like we can get more benefits from the ability to compare multiple intetgers in one operations than from summing up two integers in one operation. Moving back to ints helps compare 2x more integers at once vs. longs.

The diff is large because of the codec dance: `Lucene912PostingsFormat` and `Lucene100Codec` moved to `lucene/backward-codecs` and a new `Lucene101PostingsFormat` is a copy of the previous `Lucene912PostingsFormat` with a move from long[] arrays to int[] arrays, and changes to the on-disk format for blocks of packed integers.

Note that `DataInput#readGroupVInt` and `VectorUtilSupport#findNextGEQ` have been cleaned up to only support `int[]` and no longer `long[]`.

### CODE REVIEW

**Comment by jpountz** [comment]
Here is a `luceneutil` run against `wikibigall`:

```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                 CountOrHighHigh       76.20      (1.0%)       74.54      (0.9%)   -2.2% (  -4% -    0%) 0.000
                  CountOrHighMed      143.11      (1.2%)      141.35      (1.0%)   -1.2% (  -3% -    0%) 0.001
                CountAndHighHigh       57.54      (1.1%)       57.30      (0.9%)   -0.4% (  -2% -    1%) 0.189
                      TermDTSort      376.04      (6.4%)      374.67      (5.7%)   -0.4% ( -11% -   12%) 0.853
                     AndHighHigh       91.53      (1.4%)       91.32      (1.9%)   -0.2% (  -3% -    3%) 0.669
           HighTermDayOfYearSort      901.22      (3.8%)      899.77      (3.4%)   -0.2% (  -7% -    7%) 0.890
                      AndHighLow     1205.47      (1.7%)     1203.76      (2.0%)   -0.1% (  -3% -    3%) 0.810
                       OrHighMed      206.20      (2.5%)      206.34      (2.7%)    0.1% (  -5% -    5%) 0.935
                    OrNotHighLow     1148.24      (2.1%)     1149.12      (2.0%)    0.1% (  -3% -    4%) 0.908
                       OrHighLow      756.64      (1.7%)      757.29      (1.6%)    0.1% (  -3% -    3%) 0.872
                         MedTerm      742.62      (2.1%)      743.99      (2.2%)    0.2% (  -4% -    4%) 0.793
                      AndHighMed      184.06      (1.4%)      184.45      (1.3%)    0.2% (  -2% -    2%) 0.622
                        PKLookup      271.47      (2.2%)      272.19      (2.5%)    0.3% (  -4% -    5%) 0.728
                      OrHighRare      280.73      (4.2%)      281.50      (5.4%)    0.3% (  -8% -   10%) 0.861
                   OrHighNotHigh      262.74      (2.8%)      263.51      (2.6%)    0.3% (  -4% -    5%) 0.736
                    AndStopWords       32.53      (4.5%)       32.64      (4.2%)    0.3% (  -8% -    9%) 0.806
                        HighTerm      443.84      (2.8%)      445.95      (2.1%)    0.5% (  -4% -    5%) 0.551
                    OrHighNotMed      485.11      (3.0%)      487.66      (3.4%)    0.5% (  -5% -    7%) 0.612
                         LowTerm     1138.20      (2.7%)     1144.70      (2.8%)    0.6% (  -4% -    6%) 0.519
                          Fuzzy2       75.60      (2.1%)       76.05      (2.5%)    0.6% (  -3% -    5%) 0.429
                        Wildcard      116.11      (3.2%)      116.86      (4.1%)    0.6% (  -6% -    8%) 0.597
                      OrHighHigh       93.59      (3.6%)       94.24      (3.4%)    0.7% (  -6% -    7%) 0.538
                   OrNotHighHigh      261.04      (2.8%)      262.93      (2.4%)    0.7% (  -4% -    6%) 0.396
                          Fuzzy1       80.27      (2.6%)       80.86      (2.6%)    0.7% (  -4% -    6%) 0.381
             And2Terms2StopWords      161.95      (2.6%)      163.18      (2.5%)    0.8% (  -4% -    5%) 0.354
            HighTermTitleBDVSort       15.67      (6.6%)       15.80      (5.8%)    0.8% ( -10% -   14%) 0.677
                    OrHighNotLow      447.52      (4.0%)      451.75      (4.1%)    0.9% (  -6% -    9%) 0.471
                       And3Terms      178.15      (3.2%)      179.88      (2.8%)    1.0% (  -4% -    7%) 0.319
              Or2Terms2StopWords      164.13      (3.7%)      166.04      (3.4%)    1.2% (  -5% -    8%) 0.312
                     OrStopWords       36.12      (6.7%)       36.55      (6.1%)    1.2% ( -10% -   14%) 0.564
                        Or3Terms      178.00      (3.7%)      180.14      (3.5%)    1.2% (  -5% -    8%) 0.309
                         Prefix3       70.94      (4.1%)       71.81      (8.1%)    1.2% ( -10% -   13%) 0.554
                          IntNRQ      179.05      (5.1%)      181.32      (5.4%)    1.3% (  -8% -   12%) 0.459
               HighTermMonthSort     3413.39      (2.2%)     3459.32      (3.0%)    1.3% (  -3% -    6%) 0.111
                    OrNotHighMed      384.09      (3.2%)      389.69      (2.5%)    1.5% (  -4% -    7%) 0.112
                          OrMany       19.16      (3.5%)       19.44      (3.6%)    1.5% (  -5% -    8%) 0.203
                       CountTerm     9388.28      (3.3%)     9587.31      (4.2%)    2.1% (  -5% -    9%) 0.082
               HighTermTitleSort      135.48      (1.9%)      139.76      (3.3%)    3.2% (  -1% -    8%) 0.000
                 CountAndHighMed      160.02      (1.3%)      168.58      (1.3%)    5.4% (   2% -    7%) 0.000
```

The `CountAndHighMed` and `HighTermTitleSort` speedups are consistently reproducible. I believe that the former is due to being able to compare 8 lanes at once instead of 4, and the latter is due to the better memory efficiency of the postings reader now that it stores two int[128] instead of two long[128] (this task creates a few tens of postings readers under the hood).

**Comment by jpountz** [comment]
I plan on merging tomorrow, so that we have two data points with longs on nightly benchmarks before seeing how it performs with ints.

**Review by gsmiller — APPROVED** [review]
Seems good. Tried to look through the actual changes (and not all the codec wiring stuff).

File: lucene/core/src/java/org/apache/lucene/internal/vectorization/PostingDecodingUtil.java [hunk 1]
```diff
@@ -55,4 +55,30 @@ public void splitLongs(
...
+   *
+   * <ul>
+   *   <li>Read {@code count} ints.
+   *   <li>For all {@code i} &gt;= 0 so that {@code bShift - i * dec} &gt; 0, apply shift {@code
+   *       bShift - i * dec} and store the result in {@code b} at offset {@code count * i}.
+   *   <li>Apply mask {@code cMask} and store the result in {@code c} starting at offset {@code
+   *       cIndex}.
+   * </ul>
+   */
+  public void splitInts(
```
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/internal/vectorization/PostingDecodingUtil.java:44]: FWIW this class may only be used from a very small set of explicitly named classes, see `org.apache.lucene.internal.vectorization.VectorizationProvider#VALID_CALLERS`, so there is no risk that users use this API.

**Comment by jpountz** [comment]
Nightly benchmarks confirmed a speedup from the combination of #13958 (SIMD for advancing within a block) and this PR. ([CountAndHighHigh](https://benchmarks.mikemccandless.com/CountAndHighHigh.html), [CountAndHighMed](https://benchmarks.mikemccandless.com/CountAndHighMed.html))

Exhaustive evaluation of disjunctive queries looks stable or barely slower ([CountOrHighHigh](https://benchmarks.mikemccandless.com/CountOrHighHigh.html), [CountOrHighMed](https://benchmarks.mikemccandless.com/CountOrHighMed.html)) while we could have expected a slowdown from the loss of some optimizations in block decoding.

**Pull request merged**